README.md
<p align="center">
  <img src="/assets/images/xcmonkey.png"/>
</p>

<p align="center">
  <a href="https://github.com/testableapple/xcmonkey/actions"><img src="https://github.com/testableapple/xcmonkey/actions/workflows/test.yml/badge.svg" /></a>
  <a href="https://sonarcloud.io/summary/new_code?id=testableapple_xcmonkey"><img src="https://sonarcloud.io/api/project_badges/measure?project=testableapple_xcmonkey&metric=coverage" /></a>
  <a href="https://rubygems.org/gems/xcmonkey"><img src="https://img.shields.io/gem/v/xcmonkey.svg?style=flat" /></a>
  <a href="/LICENSE"><img src="https://img.shields.io/badge/license-MIT-green.svg?style=flat" /></a>
</p>

## Description

*xcmonkey* is a tool for doing stress testing of iOS apps. It's inspired by and has similar goals to [*monkey*](https://developer.android.com/studio/test/monkey) on Android.

Under the hood, *xcmonkey* uses [iOS Development Bridge](https://fbidb.io/) as a driver, that's why it's pretty smart and can do a lot of things, such as taps, swipes and presses. All that comes «pseudo-random» because it has access to the screen hierarchy, and so can either do actions blindly (like tapping on random points) or precisely (like tapping on the existing elements).

## Prerequisites

```bash
brew install facebook/fb/idb-companion
pip3.6 install fb-idb
```

## Installation

```bash
gem install xcmonkey
```

If you prefer to use [*bundler*](https://bundler.io/), add the following line to your `Gemfile`:

```ruby
gem 'xcmonkey'
```

## Usage

### To run a stress test

```bash
$ xcmonkey test --event-count 100 --bundle-id "com.apple.Maps" --udid "413EA256-CFFB-4312-94A6-12592BEE4CBA"

12:44:19.343: Device info: {
  "name": "iPhone 14 Pro",
  "udid": "413EA256-CFFB-4312-94A6-12592BEE4CBA",
  "state": "Booted",
  "type": "simulator",
  "os_version": "iOS 16.2",
  "architecture": "x86_64",
  "path": "/tmp/idb/413EA256-CFFB-4312-94A6-12592BEE4CBA_companion.sock",
  "is_local": true,
  "companion": "/tmp/idb/413EA256-CFFB-4312-94A6-12592BEE4CBA_companion.sock"
}

12:44:22.550: App info: {
  "bundle_id": "com.apple.Maps",
  "name": "Maps",
  "install_type": "system",
  "architectures": [
    "x86_64",
    "arm64"
  ],
  "process_state": "Running",
  "debuggable": false,
  "pid": "49186"
}

12:44:23.203: Tap: {
  "x": 53,
  "y": 749
}

12:44:23.511: Swipe (0.5s): {
  "x": 196,
  "y": 426
} => {
  "x": 143,
  "y": 447
}

12:44:24.355: Press (1.2s): {
  "x": 143,
  "y": 323
}
```

### To repeat the stress test from generated session

```bash
xcmonkey repeat --session-path "./xcmonkey-session.json"
```

### To describe the required point

```bash
xcmonkey describe -x 20 -y 625 --udid "413EA256-CFFB-4312-94A6-12592BEE4CBA"
```

### Test options reference

The table below lists all options you can include on the `xcmonkey test` command line.

| Category | Option | Description | Default |
| --- | --- | --- | --- |
| **General** | `-h, --help` | Display help documentation | |
| | `-v, --version` | Display version information | |
| | `-t, --trace` | Display backtrace when an error occurs | |
| **Events** | `-u, --udid <string>` | Set device UDID | |
| | `-b, --bundle-id <string>` | Set target bundle identifier | |
| | `-s, --session-path <string>` | Path where test session should be saved | |
| | `-e, --event-count <integer>` | Set events count | `60` |
| | `--exclude-taps` | Exclude taps from gestures list | `false` |
| | `--exclude-swipes` | Exclude swipes from gestures list | `false` |
| | `--exclude-presses` | Exclude presses from gestures list | `false` |
| | `--disable-simulator-keyboard` | Disable simulator keyboard | `false` |
| **Debugging** | `--ignore-crashes` | Ignore app crashes | `false` |
| | `--throttle <milliseconds>` | Fixed delay between events | `0` |

## [fastlane](https://github.com/fastlane/fastlane) integration

To get started with [xcmonkey fastlane plugin](https://github.com/testableapple/fastlane-plugin-xcmonkey), add it to your project by running:

```bash
fastlane add_plugin xcmonkey
```

### Usage

```ruby
  bundle_id = 'com.apple.Maps'
  device = 'iPhone 14'
  sim = FastlaneCore::Simulator.all.filter { |d| d.name == device }.max_by(&:os_version)
  udid = sim.udid

  xcmonkey(udid: udid, bundle_id: bundle_id)
end
```

## Code of Conduct

Help us keep *xcmonkey* open and inclusive. Please read and follow our [Code of Conduct](CODE_OF_CONDUCT.md).

## License

This project is licensed under the terms of the MIT license. See the [LICENSE](LICENSE) file.


CHANGELOG.md
# xcmonkey changelog

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).

## [1.3.1](https://github.com/testableapple/xcmonkey/releases/tag/1.3.1)

_February 08, 2023_

### ✅ Added

- Terminate app after test

## [1.3.0](https://github.com/testableapple/xcmonkey/releases/tag/1.3.0)

_January 29, 2023_

### 🔄 Changed

- `duration` option is renamed to `event-count`

### ✅ Added

- New test options that allow to
  - specify gestures to exclude from the test
  - set up throttle between events
  - ignore app crashes

## [1.2.0](https://github.com/testableapple/xcmonkey/releases/tag/1.2.0)

_January 23, 2023_

### ✅ Added

- Force keep xcmonkey in the target app

## [1.1.0](https://github.com/testableapple/xcmonkey/releases/tag/1.1.0)

_January 20, 2023_

### ✅ Added

- Support [fastlane](https://github.com/fastlane/fastlane) integration

## [1.0.0](https://github.com/testableapple/xcmonkey/releases/tag/1.0.0)

_January 09, 2023_

### ✅ Added

- xcmonkey `test`, `repeat` and `describe` commands


spec/driver_spec.rb
describe Driver do
  let(:udid) { `xcrun simctl list | grep " iPhone 14 Pro Max"`.split("\n")[0].split('(')[1].split(')')[0] }
  let(:bundle_id) { 'com.apple.Maps' }
  let(:driver) { described_class.new(udid: udid, bundle_id: bundle_id) }
  let(:driver_with_session) { described_class.new(udid: udid, bundle_id: bundle_id, session_actions: [{ type: 'tap', x: 0, y: 0 }]) }

  before do
    allow(Logger).to receive(:info)
  end

  it 'verifies that sumulator was booted' do
    error_message = "Failed to boot #{udid}"
    expect(Logger).not_to receive(:error).with(error_message, payload: nil)
    expect { driver.boot_simulator }.not_to raise_error
  end

  it 'verifies that ui can be described' do
    driver.boot_simulator
    ui = driver.describe_ui
    expect(ui).not_to be_empty
  end

  it 'verifies that list of targets can be showed' do
    list_targets = driver.list_targets
    expect(list_targets).not_to be_empty
  end

  it 'verifies that list of apps can be showed' do
    driver.boot_simulator
    app_exists = driver.list_apps.any? { |app| app['bundle_id'] == bundle_id }
    expect(app_exists).to be(true)
  end

  it 'verifies that app installed' do
    driver.boot_simulator
    error_message = "App #{bundle_id} is not installed on device #{udid}"
    expect(Logger).not_to receive(:error).with(error_message, payload: nil)
    expect { driver.ensure_app_installed }.not_to raise_error
  end

  it 'verifies that app is not installed' do
    driver.boot_simulator
    bundle_id = 'fake.app.bundle.id'
    error_message = "App #{bundle_id} is not installed on device #{udid}"
    driver = described_class.new(udid: udid, bundle_id: bundle_id)
    expect(Logger).to receive(:log).with(error_message, color: :light_red, payload: nil)
    expect { driver.ensure_app_installed }.to raise_error(SystemExit) { |e| expect(e.status).to eq(1) }
  end

  it 'verifies that device exists' do
    expect(Logger).not_to receive(:error)
    expect(driver).to receive(:boot_simulator)
    expect(driver).to receive(:configure_simulator_keyboard)
    expect { driver.ensure_device_exists }.not_to raise_error
  end

  it 'verifies that device does not exist' do
    udid = '1234-5678'
    error_message = "Can't find device #{udid}"
    driver = described_class.new(udid: udid, bundle_id: bundle_id)
    expect(Logger).to receive(:log).with(error_message, color: :light_red, payload: nil)
    expect { driver.ensure_device_exists }.to raise_error(SystemExit) { |e| expect(e.status).to eq(1) }
  end

  it 'verifies that idb installed' do
    error_message = "'idb' doesn't seem to be installed"
    expect(Logger).not_to receive(:error).with(error_message, payload: nil)
    expect { driver.send(:ensure_driver_installed) }.not_to raise_error
  end

  it 'verifies that element central coordinates can be found' do
    element = JSON.parse('{ "frame": { "x": 0, "y": 0, "width": 100, "height": 100 } }')
    expected_coordinates = { 'x': 50, 'y': 50 }
    actual_coordinates = driver.central_coordinates(element)
    expect(actual_coordinates).to eq(expected_coordinates)
  end

  it 'verifies that device info can be for booted simulator' do
    driver.boot_simulator
    expect(driver.device_info).not_to be_empty
  end

  it 'verifies that device info can be for not booted simulator' do
    driver.shutdown_simulator
    expect(driver.device_info).not_to be_empty
  end

  it 'verifies that screen size can be found' do
    driver.boot_simulator
    screen_size = driver.screen_size
    expect(screen_size[:width]).to be > 0
    expect(screen_size[:height]).to be > 0
  end

  it 'verifies that random coordinates can be found' do
    driver.boot_simulator
    coordinates = driver.random_coordinates
    expect(coordinates[:x]).to be > 0
    expect(coordinates[:y]).to be > 0
  end

  it 'verifies swipe duration' do
    expect(driver.swipe_duration).to be_between(0.1, 0.7)
  end

  it 'verifies press duration' do
    expect(driver.press_duration).to be_between(0.5, 1.5)
  end

  it 'verifies that simulator keyboard can be enabled' do
    allow(driver).to receive(:is_simulator_keyboard_enabled?).and_return(false)
    driver = described_class.new(udid: udid, bundle_id: bundle_id, disable_simulator_keyboard: false)
    expect(driver).to receive(:shutdown_simulator)
    driver.configure_simulator_keyboard
    keyboard_state = `defaults read com.apple.iphonesimulator`.split("\n").grep(/ConnectHardwareKeyboard/)
    expect(keyboard_state).not_to be_empty
    expect(keyboard_state.first).to include('0')
  end

  it 'verifies that simulator keyboard can be disabled' do
    allow(driver).to receive(:is_simulator_keyboard_enabled?).and_return(true)
    driver = described_class.new(udid: udid, bundle_id: bundle_id, disable_simulator_keyboard: true)
    expect(driver).to receive(:shutdown_simulator)
    driver.configure_simulator_keyboard
    keyboard_state = `defaults read com.apple.iphonesimulator`.split("\n").grep(/ConnectHardwareKeyboard/)
    expect(keyboard_state).not_to be_empty
    expect(keyboard_state.first).to include('1')
  end

  it 'verifies that app can be launched with waiting' do
    expect(Logger).not_to receive(:error)
    expect(driver).to receive(:wait_until_app_launched)
    driver.boot_simulator
    driver.terminate_app(bundle_id)
    expect { driver.launch_app(target_bundle_id: bundle_id, wait_for_state_update: true) }.not_to raise_error
  end

  it 'verifies that app can be launched without waiting' do
    expect(Logger).not_to receive(:error)
    expect(driver).not_to receive(:wait_until_app_launched)
    driver.boot_simulator
    driver.terminate_app(bundle_id)
    expect { driver.launch_app(target_bundle_id: bundle_id) }.not_to raise_error
  end

  it 'verifies tap in new session' do
    driver.boot_simulator
    coordinates = { x: 1, y: 1 }
    driver.tap(coordinates: coordinates)
    expect(driver.instance_variable_get(:@session)[:actions]).not_to be_empty
  end

  it 'verifies tap in old session' do
    driver_with_session.boot_simulator
    coordinates = { x: 1, y: 1 }
    driver_with_session.tap(coordinates: coordinates)
    expect(driver_with_session.instance_variable_get(:@session)[:actions]).to be_empty
  end

  it 'verifies press in new session' do
    driver.boot_simulator
    duration = 0.5
    coordinates = { x: 1, y: 1 }
    driver.press(coordinates: coordinates, duration: duration)
    expect(driver.instance_variable_get(:@session)[:actions]).not_to be_empty
  end

  it 'verifies press in old session' do
    driver_with_session.boot_simulator
    duration = 0.5
    coordinates = { x: 1, y: 1 }
    driver_with_session.press(coordinates: coordinates, duration: duration)
    expect(driver_with_session.instance_variable_get(:@session)[:actions]).to be_empty
  end

  it 'verifies swipe in new session' do
    driver.boot_simulator
    duration = 0.5
    start_coordinates = { x: 1, y: 1 }
    end_coordinates = { x: 2, y: 2 }
    driver.swipe(start_coordinates: start_coordinates, end_coordinates: end_coordinates, duration: duration)
    expect(driver.instance_variable_get(:@session)[:actions]).not_to be_empty
  end

  it 'verifies swipe in old session' do
    driver_with_session.boot_simulator
    duration = 0.5
    start_coordinates = { x: 1, y: 1 }
    end_coordinates = { x: 2, y: 2 }
    driver_with_session.swipe(start_coordinates: start_coordinates, end_coordinates: end_coordinates, duration: duration)
    expect(driver_with_session.instance_variable_get(:@session)[:actions]).to be_empty
  end

  it 'verifies that session can be saved' do
    driver = described_class.new(udid: udid, bundle_id: bundle_id, session_path: Dir.pwd)
    expect(File).to receive(:write)
    driver.instance_variable_set(:@session, { params: {}, actions: [] })
    driver.save_session
  end

  it "verifies that session won't be saved if path is not provided" do
    expect(File).not_to receive(:write)
    driver.save_session
  end

  it 'verifies that monkey_test_precondition works fine' do
    driver.monkey_test_precondition
    app_info = driver.list_apps.detect { |app| app['bundle_id'] == bundle_id }
    app_is_running = app_info && app_info['process_state'] == 'Running'
    expect(app_is_running).to be(true)
    expect(driver.instance_variable_get(:@running_apps)).not_to be_nil
  end

  it 'verifies that monkey_test works fine' do
    params = { udid: udid, bundle_id: bundle_id, event_count: 1, session_path: Dir.pwd }
    driver = described_class.new(params)
    driver.monkey_test(Xcmonkey.new(params).gestures)
    expect(driver.instance_variable_get(:@session)[:actions]).not_to be_empty
  end

  it 'verifies that repeat_monkey_test works fine' do
    session_actions = [
      { 'type' => 'tap', 'x' => 10, 'y' => 10 },
      { 'type' => 'press', 'x' => 11, 'y' => 11, 'duration' => 1.4 },
      { 'type' => 'swipe', 'x' => 12, 'y' => 12, 'endX' => 15, 'endY' => 15, 'duration' => 0.3 }
    ]
    driver = described_class.new(udid: udid, bundle_id: bundle_id, session_actions: session_actions)
    expect(driver).to receive(:tap).with(coordinates: { x: 10, y: 10 })
    expect(driver).to receive(:press).with(coordinates: { x: 11, y: 11 }, duration: 1.4)
    expect(driver).to receive(:swipe).with(start_coordinates: { x: 12, y: 12 }, end_coordinates: { x: 15, y: 15 }, duration: 0.3)
    driver.repeat_monkey_test
    expect(driver.instance_variable_get(:@session)[:actions]).to be_empty
  end

  it 'verifies that unknown actions does not break repeat_monkey_test' do
    driver = described_class.new(udid: udid, bundle_id: bundle_id, session_actions: [{ 'type' => 'test', 'x' => 10, 'y' => 10 }])
    expect(driver).to receive(:monkey_test_precondition)
    expect(driver).not_to receive(:tap)
    expect(driver).not_to receive(:press)
    expect(driver).not_to receive(:swipe)
    driver.repeat_monkey_test
    expect(driver.instance_variable_get(:@session)[:actions]).to be_empty
  end

  it 'verifies that running apps are tracked' do
    new_app_bundle_id = 'com.apple.Preferences'
    driver.terminate_app(new_app_bundle_id)
    driver.monkey_test_precondition
    driver.launch_app(target_bundle_id: new_app_bundle_id, wait_for_state_update: true)
    expect(driver).to receive(:launch_app).with(target_bundle_id: bundle_id)
    expect(driver).to receive(:terminate_app).with(new_app_bundle_id)
    driver.track_running_apps
  end

  it 'verifies that running apps can be determined' do
    driver.terminate_app(bundle_id)
    sum = driver.list_running_apps.size
    driver.launch_app(target_bundle_id: bundle_id)
    expect(driver.list_running_apps.size).to eq(sum + 1)
  end

  it 'verifies that app state change can be determined' do
    driver.launch_app(target_bundle_id: bundle_id)
    allow(driver).to receive(:detect_app_in_background).and_return(true)
    expect(driver).not_to receive(:save_session)
    expect(driver).to receive(:launch_app)
    expect { driver.detect_app_state_change }.not_to raise_error
  end

  it 'verifies that background is the invalid app state' do
    driver.terminate_app(bundle_id)
    expect(driver).to receive(:save_session)
    expect { driver.detect_app_state_change }.to raise_error(SystemExit) { |e| expect(e.status).to eq(1) }
  end

  it 'verifies that foreground is the valid app state' do
    driver.launch_app(target_bundle_id: bundle_id, wait_for_state_update: true)
    expect { driver.detect_app_state_change }.not_to raise_error
  end

  it 'verifies that app crashes can be ignored' do
    driver = described_class.new(udid: udid, bundle_id: bundle_id, session_path: Dir.pwd, ignore_crashes: true)
    driver.terminate_app(bundle_id)
    expect(driver).not_to receive(:save_session)
    expect(driver).to receive(:launch_app)
    expect { driver.detect_app_state_change }.not_to raise_error
  end

  it 'verifies that background state can be determined' do
    driver.terminate_app(bundle_id)
    expect(driver.detect_app_in_background).to be(true)
  end

  it 'verifies that xcmonkey behaves as expected on real devices' do
    udid = '1234-5678'
    driver = described_class.new(udid: udid, bundle_id: bundle_id)
    allow(driver).to receive(:list_targets).and_return([{ 'udid' => udid, 'type' => 'device' }])
    expect { driver.ensure_device_exists }.to raise_error(SystemExit) { |e| expect(e.status).to eq(1) }
  end

  it 'verifies that test can be slowed down' do
    throttle = 1000
    driver = described_class.new(udid: udid, bundle_id: bundle_id, session_path: Dir.pwd, throttle: throttle)
    expect(driver).to receive(:sleep).with(throttle / 1000.0)
    driver.check_speed_limit
  end

  it 'verifies that test ignores throttle by default' do
    expect(driver).not_to receive(:sleep)
    driver.check_speed_limit
  end

  it 'verifies that running apps are tracked on second entry with throttle' do
    driver = described_class.new(udid: udid, bundle_id: bundle_id, session_path: Dir.pwd, throttle: 1)
    driver.launch_app(target_bundle_id: bundle_id)
    expect(driver).to receive(:track_running_apps)
    driver.checkup(1)
  end

  it 'verifies that running apps are not tracked on second entry without throttle' do
    driver.launch_app(target_bundle_id: bundle_id)
    expect(driver).not_to receive(:track_running_apps)
    driver.checkup(1)
  end

  it 'verifies that simulator was not booted' do
    driver.shutdown_simulator
    error_message = "Failed to boot #{udid}"
    allow(driver).to receive(:device_info).and_return({ 'state' => 'Unknown' })
    expect(Logger).to receive(:log).with(error_message, color: :light_red, payload: nil)
    expect { driver.boot_simulator }.to raise_error(SystemExit) { |e| expect(e.status).to eq(1) }
  end
end


spec/logger_spec.rb
describe Logger do
  let(:message) { 'test' }
  let(:payload) { '{"test": true}' }

  it 'verifies info log without payload' do
    expect(described_class).to receive(:log).with(message, color: :light_cyan, payload: nil)
    described_class.info(message)
  end

  it 'verifies info log with payload' do
    expect(described_class).to receive(:log).with(message, color: :light_cyan, payload: payload)
    described_class.info(message, payload: payload)
  end

  it 'verifies warning log without payload' do
    expect(described_class).to receive(:log).with(message, color: :light_yellow, payload: nil)
    described_class.warn(message)
  end

  it 'verifies warning log with payload' do
    expect(described_class).to receive(:log).with(message, color: :light_yellow, payload: payload)
    described_class.warn(message, payload: payload)
  end

  it 'verifies error log without payload' do
    expect(described_class).to receive(:log).with(message, color: :light_red, payload: nil)
    expect { described_class.error(message) }
      .to raise_error(SystemExit) { |e| expect(e.status).to eq(1) }
  end

  it 'verifies error log with payload' do
    expect(described_class).to receive(:log).with(message, color: :light_red, payload: payload)
    expect { described_class.error(message, payload: payload) }
      .to raise_error(SystemExit) { |e| expect(e.status).to eq(1) }
  end

  it 'verifies custom log without payload' do
    color = :blue
    time = Time.now
    allow(Time).to receive(:now).and_return(time)
    expected_output = "#{time.strftime('%k:%M:%S.%L')}: #{message}".colorize(color)
    expect do
      described_class.log(message, color: color, payload: nil)
    end.to output("#{expected_output}\n\n").to_stdout
  end

  it 'verifies custom log with payload' do
    color = :blue
    time = Time.now
    allow(Time).to receive(:now).and_return(time)
    expected_output = "#{time.strftime('%k:%M:%S.%L')}: #{message}".colorize(color)
    expect do
      described_class.log(message, color: color, payload: payload)
    end.to output("#{expected_output} #{payload.colorize(:light_green)}\n\n").to_stdout
  end
end


spec/spec_helper.rb
$LOAD_PATH.unshift(File.expand_path('../lib', __dir__))

require 'simplecov'

# SimpleCov.minimum_coverage(80)
SimpleCov.start

# This module is only used to check the environment is currently a testing env
module SpecHelper
end

require 'xcmonkey'


lib/xcmonkey.rb
require 'json'
require 'colorize'
require_relative 'xcmonkey/describer'
require_relative 'xcmonkey/repeater'
require_relative 'xcmonkey/version'
require_relative 'xcmonkey/logger'
require_relative 'xcmonkey/driver'

class Xcmonkey
  attr_accessor :params, :driver

  def initialize(params)
    params[:event_count] = 60 if params[:event_count].nil?
    params[:ignore_crashes] = false if params[:ignore_crashes].nil?
    params[:disable_simulator_keyboard] = false if params[:disable_simulator_keyboard].nil?
    self.params = params
    self.driver = Driver.new(params)
    ensure_required_params
  end

  def run
    driver.monkey_test(gestures)
  end

  def gestures
    taps = params[:exclude_taps] ? [] : [:precise_tap, :blind_tap] * 10
    swipes = params[:exclude_swipes] ? [] : [:precise_swipe, :blind_swipe] * 5
    presses = params[:exclude_presses] ? [] : [:precise_press, :blind_press]
    taps + swipes + presses
  end

  def ensure_required_params
    Logger.error('UDID should be provided') if params[:udid].nil?

    Logger.error('Bundle identifier should be provided') if params[:bundle_id].nil?

    Logger.error('Session path should be a directory') if params[:session_path] && !File.directory?(params[:session_path])

    if params[:event_count].nil? || !params[:event_count].kind_of?(Integer) || !params[:event_count].positive?
      Logger.error('Event count must be Integer and not less than 1')
    end
  end
end


lib/xcmonkey/repeater.rb
class Repeater
  attr_accessor :udid, :bundle_id, :disable_simulator_keyboard, :ignore_crashes, :actions, :throttle

  def initialize(params)
    validate_session(params[:session_path])
  end

  def run
    params = {
      udid: udid,
      throttle: throttle,
      bundle_id: bundle_id,
      ignore_crashes: ignore_crashes,
      disable_simulator_keyboard: disable_simulator_keyboard,
      session_actions: actions
    }
    Driver.new(params).repeat_monkey_test
  end

  def validate_session(session_path)
    Logger.error("Provided session can't be found: #{session_path}") unless File.exist?(session_path)

    session = JSON.parse(File.read(session_path))

    if session['params'].nil?
      Logger.error('Provided session is not valid: `params` should not be `nil`')
      return
    end

    self.actions = session['actions']
    Logger.error('Provided session is not valid: `actions` should not be `nil` or `empty`') if actions.nil? || actions.empty?

    self.udid = session['params']['udid']
    Logger.error('Provided session is not valid: `udid` should not be `nil`') if udid.nil?

    self.bundle_id = session['params']['bundle_id']
    Logger.error('Provided session is not valid: `bundle_id` should not be `nil`') if bundle_id.nil?

    self.throttle = session['params']['throttle']

    self.ignore_crashes = session['params']['ignore_crashes']

    self.disable_simulator_keyboard = session['params']['disable_simulator_keyboard']
  end
end


lib/xcmonkey/driver.rb
class Driver
  attr_accessor :udid, :bundle_id, :disable_simulator_keyboard, :event_count, :session_path, :session_actions, :ignore_crashes, :throttle

  def initialize(params)
    self.udid = params[:udid]
    self.throttle = params[:throttle]
    self.bundle_id = params[:bundle_id]
    self.event_count = params[:event_count]
    self.session_path = params[:session_path]
    self.ignore_crashes = params[:ignore_crashes]
    self.disable_simulator_keyboard = params[:disable_simulator_keyboard]
    self.session_actions = params[:session_actions]
    @session = { params: params, actions: [] }
    ensure_driver_installed
  end

  def monkey_test_precondition
    puts
    ensure_device_exists
    ensure_app_installed
    terminate_app(bundle_id)
    launch_app(target_bundle_id: bundle_id, wait_for_state_update: true)
    @running_apps = list_running_apps
  end

  def monkey_test(gestures)
    monkey_test_precondition
    event_count.times do |counter|
      app_elements = describe_ui.shuffle
      el1_coordinates = central_coordinates(app_elements.first)
      el2_coordinates = central_coordinates(app_elements.last)
      case gestures.sample
      when :precise_tap
        tap(coordinates: el1_coordinates)
      when :blind_tap
        tap(coordinates: random_coordinates)
      when :precise_press
        press(coordinates: el1_coordinates, duration: press_duration)
      when :blind_press
        press(coordinates: random_coordinates, duration: press_duration)
      when :precise_swipe
        swipe(
          start_coordinates: el1_coordinates,
          end_coordinates: el2_coordinates,
          duration: swipe_duration
        )
      when :blind_swipe
        swipe(
          start_coordinates: random_coordinates,
          end_coordinates: random_coordinates,
          duration: swipe_duration
        )
      else
        next
      end
      checkup(counter)
    end
    save_session
    terminate_app(bundle_id)
  end

  def repeat_monkey_test
    monkey_test_precondition
    session_actions.each_with_index do |action, counter|
      case action['type']
      when 'tap'
        tap(coordinates: { x: action['x'], y: action['y'] })
      when 'press'
        press(coordinates: { x: action['x'], y: action['y'] }, duration: action['duration'])
      when 'swipe'
        swipe(
          start_coordinates: { x: action['x'], y: action['y'] },
          end_coordinates: { x: action['endX'], y: action['endY'] },
          duration: action['duration']
        )
      else
        next
      end
      checkup(counter)
    end
    terminate_app(bundle_id)
  end

  def checkup(counter)
    detect_app_state_change
    if counter % 5 == 0 || throttle.to_i > 0 # Track running apps after every 5th action
      track_running_apps                     # (unless `throttle` was provided) to speed up the test
    end
    check_speed_limit
  end

  def describe_ui
    JSON.parse(`idb ui describe-all --udid #{udid}`)
  end

  def describe_point(x, y)
    point_info = JSON.parse(`idb ui describe-point --udid #{udid} #{x} #{y}`)
    Logger.info("x:#{x} y:#{y} point info:", payload: JSON.pretty_generate(point_info))
    point_info
  end

  def launch_app(target_bundle_id:, wait_for_state_update: false)
    `idb launch --udid #{udid} #{target_bundle_id}`
    wait_until_app_launched(target_bundle_id) if wait_for_state_update
  end

  def terminate_app(target_bundle_id)
    `idb terminate --udid #{udid} #{target_bundle_id} 2>/dev/null`
  end

  def boot_simulator
    `idb boot #{udid}`
    Logger.error("Failed to boot #{udid}") if device_info['state'] != 'Booted'
  end

  def shutdown_simulator
    `idb shutdown #{udid}`
  end

  def configure_simulator_keyboard
    shutdown_simulator
    keyboard_status = disable_simulator_keyboard ? 1 : 0
    `defaults write com.apple.iphonesimulator ConnectHardwareKeyboard #{keyboard_status}`
  end

  def list_targets
    @targets ||= `idb list-targets --json`.split("\n").map! { |target| JSON.parse(target) }
    @targets
  end

  def list_apps
    `idb list-apps --udid #{udid} --json`.split("\n").map! { |app| JSON.parse(app) }
  end

  def list_running_apps
    list_apps.select { |app| app['process_state'] == 'Running' }
  end

  def ensure_app_installed
    return if list_apps.any? { |app| app['bundle_id'] == bundle_id }

    Logger.error("App #{bundle_id} is not installed on device #{udid}")
  end

  def ensure_device_exists
    device = list_targets.detect { |target| target['udid'] == udid }
    Logger.error("Can't find device #{udid}") if device.nil?

    Logger.info('Device info:', payload: JSON.pretty_generate(device))
    if device['type'] == 'simulator'
      configure_simulator_keyboard
      boot_simulator
    else
      Logger.error('xcmonkey does not support real devices yet. ' \
                   'For more information see https://github.com/testableapple/xcmonkey/issues/7')
    end
  end

  def tap(coordinates:)
    Logger.info('Tap:', payload: JSON.pretty_generate(coordinates))
    @session[:actions] << { type: :tap, x: coordinates[:x], y: coordinates[:y] } unless session_actions
    `idb ui tap --udid #{udid} #{coordinates[:x]} #{coordinates[:y]}`
  end

  def press(coordinates:, duration:)
    Logger.info("Press (#{duration}s):", payload: JSON.pretty_generate(coordinates))
    @session[:actions] << { type: :press, x: coordinates[:x], y: coordinates[:y], duration: duration } unless session_actions
    `idb ui tap --udid #{udid} --duration #{duration} #{coordinates[:x]} #{coordinates[:y]}`
  end

  def swipe(start_coordinates:, end_coordinates:, duration:)
    payload = "#{JSON.pretty_generate(start_coordinates)} => #{JSON.pretty_generate(end_coordinates)}"
    Logger.info("Swipe (#{duration}s):", payload: payload)
    unless session_actions
      @session[:actions] << {
        type: :swipe,
        x: start_coordinates[:x],
        y: start_coordinates[:y],
        endX: end_coordinates[:x],
        endY: end_coordinates[:y],
        duration: duration
      }
    end
    coordinates = "#{start_coordinates[:x]} #{start_coordinates[:y]} #{end_coordinates[:x]} #{end_coordinates[:y]}"
    `idb ui swipe --udid #{udid} --duration #{duration} #{coordinates}`
  end

  def central_coordinates(element)
    frame = element['frame']
    x = (frame['x'] + (frame['width'] / 2)).abs.to_i
    y = (frame['y'] + (frame['height'] / 2)).abs.to_i
    {
      x: x > screen_size[:width].to_i ? random_coordinates[:x] : x,
      y: y > screen_size[:height].to_i ? random_coordinates[:y] : y
    }
  end

  def random_coordinates
    {
      x: rand(0..screen_size[:width].to_i),
      y: rand(0..screen_size[:height].to_i)
    }
  end

  def device_info
    @device_info ||= JSON.parse(`idb describe --udid #{udid} --json`)
    @device_info
  end

  def screen_size
    screen_dimensions = device_info['screen_dimensions']
    {
      width: screen_dimensions['width_points'],
      height: screen_dimensions['height_points']
    }
  end

  def swipe_duration
    rand(0.1..0.7).ceil(1)
  end

  def press_duration
    rand(0.5..1.5).ceil(1)
  end

  def save_session
    return if session_path.nil?

    File.write("#{session_path}/xcmonkey-session.json", JSON.pretty_generate(@session))
  end

  # This function takes ≈200ms
  def track_running_apps
    current_list_of_running_apps = list_running_apps
    if @running_apps != current_list_of_running_apps
      currently_running_bundle_ids = current_list_of_running_apps.map { |app| app['bundle_id'] }
      previously_running_bundle_ids = @running_apps.map { |app| app['bundle_id'] }
      new_apps = currently_running_bundle_ids - previously_running_bundle_ids

      return if new_apps.empty?

      launch_app(target_bundle_id: bundle_id)

      new_apps.each do |id|
        Logger.warn("Shutting down: #{id}")
        terminate_app(id)
      end
    end
  end

  # This function takes ≈300ms
  def detect_app_state_change
    return unless detect_app_in_background

    target_app_is_running = list_running_apps.any? { |app| app['bundle_id'] == bundle_id }

    if target_app_is_running || ignore_crashes
      launch_app(target_bundle_id: bundle_id)
    else
      save_session
      Logger.error("Target app has crashed or been terminated")
    end
  end

  def detect_app_in_background
    current_app_label = describe_ui.detect { |el| el['type'] == 'Application' }['AXLabel']
    current_app_label.nil? || current_app_label.strip.empty?
  end

  def check_speed_limit
    sleep(throttle / 1000.0) if throttle.to_i > 0
  end

  private

  def ensure_driver_installed
    Logger.error("'idb' doesn't seem to be installed") if `which idb`.strip.empty?
  end

  def wait_until_app_launched(target_bundle_id)
    app_is_running = false
    current_time = Time.now
    while !app_is_running && Time.now < current_time + 5
      app_info = list_apps.detect { |app| app['bundle_id'] == target_bundle_id }
      app_is_running = app_info && app_info['process_state'] == 'Running'
    end
    Logger.error("Can't run the app #{target_bundle_id}") unless app_is_running
    Logger.info('App info:', payload: JSON.pretty_generate(app_info))
  end
end
